acme/dns_challenge.go

```diff
@@ -60,50 +61,129 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 		}
 	}()
 
-	fqdn, _, _ := DNS01Record(domain, keyAuth)
+	fqdn, value, _ := DNS01Record(domain, keyAuth)
 
-	preCheckDNS(domain, fqdn)
+	err = waitFor(60, 2, func() (bool, error) {
+		return preCheckDNS(domain, fqdn, value)
+	})
 
 	return s.validate(s.jws, domain, chlng.URI, challenge{Resource: "challenge", Type: chlng.Type, Token: chlng.Token, KeyAuthorization: keyAuth})
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
+// checkDnsPropagation queries all nameservers that are authoritative for 'domain'
+// to check whether DNS propagation for the expected DNS record is complete.
+func checkDnsPropagation(domain, fqdn, value string) (bool, error) {
+	nameservers, err := lookupNameservers(toFqdn(domain))
```

Comment:
I am dissatisfied with this explanation for three reasons.

First, by looking up nameservers for `domain` instead of `fqdn`, `lego` will be checking for the `_acme-challenge IN TXT` record differently than normal recursive resolvers, which would follow such a delegation. This includes `boulder`, which [implements DNS01](https://github.com/letsencrypt/boulder/blob/e23cf3991ec298e9330a87394b82cceed862ffb5/va/validation-authority.go#L432-L434) by [asking a normal resolver to perform a normal recursive lookup](https://github.com/letsencrypt/boulder/blob/0de0ed550b9869792619363586af04cd2873e848/bdns/dns.go#L250-L251).

Second, we agree that the delegation `_acme-challenge.foo.bar.com IN NS ns.confusion.party` and subsequent creation of a `TXT` record at `ns.confusion.party` would not indicate that I have authority over `bar.com`, but I never said it did. Instead, I argue that this record specifically means that `bar.com` has *delegated the authority* to issue ACME certificates for `foo.bar.com` to `ns.confusion.party`. If I can fulfill a DNS-01 challenge by adding a record at `ns.confusion.party`, then I've fulfillled the DNS-01 challenge, full stop.

There are situations where it is undesirable or impractical from a policy/procedure standpoint to modify `bar.com` within short periods of time. (Think of those big companies with those big meetings for change control that needs oodles of review.) By delegating `_acme-challenge` ahead of time to a different nameserver with different controls, it's possible to satisfy `boulder` with a cron job intended to renew certificates, and it's possible to satisfy The Powers That Be™ by eliminating the need to dynamically modify `bar.com` in any way. I believe `lego` should not object.

Finally, the ACME spec does not require that `_acme-challenge.foo.bar.com` be served by the same nameservers as `foo.bar.com`, only that [a query for TXT records under the validation domain name – `_acme-challenge.*` – returns a matching value](https://tools.ietf.org/html/draft-ietf-acme-acme-01#section-7.5).